source/slang/slang-ir-inst-defs.h

```diff
@@ -166,7 +166,14 @@ INST(Nop, nop, 0, 0)
 INST(StructType, struct, 0, PARENT)
 INST(InterfaceType, interface, 0, PARENT)
 
-INST_RANGE(Type, VoidType, InterfaceType)
+// A TypeType-typed IRValue represents a IRType.
+// It is used to represent a type parameter/argument in a generics.
+INST(TypeType, type_t, 0, 0)
```

Comment:
There is already a `TypeKind` instruction, so that having both `TypeType` and `TypeKind` could potentially be confusing. There are really two main options here, and we need to think carefully about which seems correct:

1. Only keep `TypeKind`, and have it so that after your transformation we may have IR functions with parameters classified by `TypeKind`. This means that the distinction between the type and value levels gets a bit muddy, since we might have value-level operations that are manipulation type-level entities.

2. Have both `TypeKind` as the classifier for generic type parameters, and `TypeType` as the classifier for value-level type parameters (with the understanding that is is really the type of RTTI/reflection information). In this case your pass would translate a generic parameter classified by `TypeKind` over to a value parameter classified by `TypeType`. In this case we might need an explicit operation to go from a type-level instruction classified by `TypeKind` to a value-level instruction classified by `TypeType`, which represents the operation of looking up RTTI/reflection information for a given type.

My gut instinct is to go for option (1) there, and only consider the more complicated option if we run into serious challenges. That would argue for *not* introducing `TypeType` and instead using the existing `TypeKind`.